storage/invoke_test.go

```diff
@@ -173,6 +171,38 @@ func TestInvoke(t *testing.T) {
 			},
 			expectFinalErr: false,
 		},
+		{
+			desc:              "non-idempotent retriable error retried when policy is RetryAlways till max-retry-count",
+			count:             4,
+			initialErr:        &googleapi.Error{Code: 500},
+			finalErr:          nil,
+			isIdempotentValue: false,
+			retry:             &retryConfig{policy: RetryAlways, maxRetryCount: 2},
+			expectFinalErr:    false,
+		},
+		{
+			desc:              "non-idempotent retriable error retried when policy is RetryNever with max-retry-count set",
+			count:             4,
+			initialErr:        &googleapi.Error{Code: 500},
+			finalErr:          nil,
+			isIdempotentValue: false,
+			retry:             &retryConfig{policy: RetryNever, maxRetryCount: 2},
+			expectFinalErr:    false,
+		},
+		{
+			desc:              "non-retriable error retried with custom fn till max-retry-count",
+			count:             4,
+			initialErr:        io.ErrNoProgress,
+			finalErr:          nil,
+			isIdempotentValue: true,
+			retry: &retryConfig{
+				shouldRetry: func(err error) bool {
+					return err == io.ErrNoProgress
+				},
+				maxRetryCount: 2,
+			},
+			expectFinalErr: false,
```

Comment:
Can we have a test case where maxRetryCount is set, but equal to `count` and `expectFinalErr` is true? This would verify that we don't get cut off early.

Reply:
Done